sunpy/net/download.py

```diff
@@ -139,9 +141,17 @@ def download(self, url, path=None, callback=None, errback=None):
         default_dir = sunpy.config.get("downloads", "download_dir")
 
         if path is None:
```

Comment:
instead of modifying this function to convert to a filename, we want to keep path as a callable, for compatibility with the VSO.

What we should change here is to allow a callable as an argument for path:

```python
if path is None:
    [do some stuff]
elif isinstance(path, six.string_types):
    path = partial(default_name, path)
elif not callable(path):
    raise ValueError("path must be: None, string or callable")
```